### PR TITLE
Bugfix 2838

### LINKED ISSUE #2838: bug: mesh sampling throws error with FLASH data
### Bug report
 
The script below attempts to create a mesh sampled particle field using a FLASH dataset with plt and part files. The expected outcome is a list of density values at the location of each particle, like those produced by the example script (https://yt-project.org/doc/reference/api/yt.data_objects.static_output.html#yt.data_objects.static_output.Dataset.add_mesh_sampling_particle_field). The script below, which is identical to the example script except with FLASH datafiles, produces the error "IndexError: index 16 is out of bounds for axis 0 with size 16". 

**Code for reproduction**
FLASH data found here:
plt file: http://use.yt/upload/a94cdffd 
part file: http://use.yt/upload/dec62ce5

test script, also copied below:
 http://paste.yt-project.org/show/266/

```python
import yt
ds = yt.load("./group_subh_hgas_iso_hdf5_plt_cnt_0000", particle_filename="./group_subh_hgas_iso_hdf5_part_0000")
ds.add_mesh_sampling_particle_field(('gas', 'density'), ptype='all')
print('The density at the location of the particle is:')
print(ds.r['all', 'cell_gas_density'])
```

**Actual outcome**

```
Traceback (most recent call last):
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 355, in _generate_particle_field
    finfo.check_available(gen_obj)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/fields/derived_field.py", line 204, in check_available
    validator(data)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/fields/derived_field.py", line 470, in __call__
    raise NeedsGridType(self.ghost_zones,self.fields)
yt.fields.field_exceptions.NeedsGridType: (0, None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "mesh_script.py", line 5, in <module>
    print(ds.r['all', 'cell_gas_density'])
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/region_expression.py", line 42, in __getitem__
    return self.all_data[item]
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 256, in __getitem__
    self.get_data(f)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 1583, in get_data
    self._generate_fields(fields_to_generate)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 1603, in _generate_fields
    fd = self._generate_field(field)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 291, in _generate_field
    tr = self._generate_particle_field(field)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 363, in _generate_particle_field
    for i, chunk in enumerate(self.chunks(field, "spatial")):
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 1477, in chunks
    self.get_data(fields)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 1583, in get_data
    self._generate_fields(fields_to_generate)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 1603, in _generate_fields
    fd = self._generate_field(field)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 291, in _generate_field
    tr = self._generate_particle_field(field)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 376, in _generate_particle_field
    rv = self.ds._get_field_info(*field)(gen_obj)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/fields/derived_field.py", line 256, in __call__
    dd = self._function(self, data)
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/geometry/grid_geometry_handler.py", line 406, in _mesh_sampling_particle_field
    return field_values[i, j, k]
  File "/XMMShared/clp3ef/yt-conda/lib/python3.7/site-packages/yt/units/yt_array.py", line 1065, in __getitem__
    ret = super(YTArray, self).__getitem__(item)
IndexError: index 16 is out of bounds for axis 0 with size 16
```

**Expected Outcome**
The example script 
```python
import yt
ds = yt.load('output_00080/info_00080.txt')
... ds.add_mesh_sampling_particle_field(('gas', 'density'), ptype='all')
print('The density at the location of the particle is:')
print(ds.r['all', 'cell_gas_density'])
```
outputs:
The density at the location of the particle is:
[9.33886124e-30 1.22174333e-28 1.20402333e-28 ... 2.77410331e-30
 8.79467609e-31 3.50665136e-30] g/cm**3

**Version Information**
  * Operating System: Linux 7
  * Python Version: 3.7.6
  * yt version: 3.6

yt installed using installation script

### PR DESCRIPTION
**Pull request by cphyc**
## PR Summary

Fixes #2838. The bug is happening when particles are found for some reason at the boundary of grids.

## PR Checklist

- [x] pass `black --check yt/`
- [x] pass `isort . --check --diff`
- [x] pass `flake8 yt/`

### CODE REVIEW

File: yt/geometry/grid_geometry_handler.py [hunk 1]
```diff
@@ -475,7 +475,16 @@ def _mesh_sampling_particle_field(field, data):
...
             i, j, k = np.floor((pos - data.LeftEdge) / data.dds).astype("int64").T
 
-            return field_values[i, j, k]
+            # Make sure all particles are within the current grid, otherwise return nan
+            maxi, maxj, maxk = field_values.shape
+
+            mask = (i < maxi) & (j < maxj) & (k < maxk)
+            mask &= (i >= 0) & (j >= 0) & (k >= 0)
+
+            result = np.full_like(i, np.nan, dtype="float64")
```
Comment (neutrinoceros) [inline yt/geometry/grid_geometry_handler.py:484]: I'm confused. Why is is `full_like(i)` ?

File: yt/geometry/grid_geometry_handler.py [hunk 2]
```diff
@@ -475,7 +475,16 @@ def _mesh_sampling_particle_field(field, data):
...
+            result[mask] = field_values[i[mask], j[mask], k[mask]]
```
Comment (neutrinoceros) [inline yt/geometry/grid_geometry_handler.py:485]: Isn't this equivalent ? (if it's not, disregard)
```suggestion
            result[mask] = field_values[i, j, k][mask]
```

Comment (cphyc) [reply yt/geometry/grid_geometry_handler.py:485]: It is not, as there may be some values of `i`, `j` or `k` that are out-of-bounds (so `field_values[i, j, k]` will fail).

File: yt/geometry/grid_geometry_handler.py (hunk 1 above)
Comment (cphyc) [reply yt/geometry/grid_geometry_handler.py:484]: I could instead use 
```suggestion
            result = np.full_like(field_value, np.nan, dtype="float64")
```

Comment (neutrinoceros) [reply yt/geometry/grid_geometry_handler.py]: Please do, it's a bit clearer :)

File: yt/geometry/grid_geometry_handler.py (hunk 2 above)
Comment (neutrinoceros) [reply yt/geometry/grid_geometry_handler.py:485]: Ok, I wasn't sure. Thanks !

File: yt/geometry/grid_geometry_handler.py [hunk 3]
```diff
@@ -475,7 +475,16 @@ def _mesh_sampling_particle_field(field, data):
...
             i, j, k = np.floor((pos - data.LeftEdge) / data.dds).astype("int64").T
 
-            return field_values[i, j, k]
+            # Make sure all particles are within the current grid, otherwise return nan
+            maxi, maxj, maxk = field_values.shape
+
+            mask = (i < maxi) & (j < maxj) & (k < maxk)
+            mask &= (i >= 0) & (j >= 0) & (k >= 0)
+
+            result = np.full_like(field_value, np.nan, dtype="float64")
```
Comment (neutrinoceros) [inline yt/geometry/grid_geometry_handler.py:484]: fix undef variable
```suggestion
            result = np.full_like(field_values, np.nan, dtype="float64")
```

**Comment by cphyc** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by cphyc** [comment]
The more I think about it, the less sure I am about this fix. In particular, it supposedly addresses the problem of particles being slightly off a grid but doesn't solve the root problem, which is why the particle was associated to this grid to start with?

**Pull request merged**